### ISSUE TITLE
Generic function decorator can't infer geneirc type of `self`

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#2802]
**Describe the bug**
A function decorator that decorates a function `f` which takes a generic first argument will fail to infer the type of the first argument of `f` if `f` is a method of a generic class.

**To Reproduce**
```python3
# pyright: strict

from typing import Callable, Generic, TypeVar

T = TypeVar("T")

def my_generic_wrapper(f: Callable[["MyClass[T]", int], str]) -> Callable[["MyClass[T]", int], str]:
    def wrapper(self: "MyClass[T]", x: int) -> str:
        return f(self, x)
    return wrapper

class MyClass(Generic[T]):
    # error: Type "MyClass[T@my_generic_wrapper]" cannot be assigned to type "MyClass[T@MyClass]"
    @my_generic_wrapper # error: 
    def do_something(self, x: int) -> str:
        return "bla"

@my_generic_wrapper # works fine
def top_level_function(c: "MyClass[str]", x: int) -> str:
    return "bla"
```

**Expected behavior**
No errors when applying `@my_generic_wrapper` to `MyClass` methods

**VS Code extension or command-line**
pyright 1.1.203

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.205, which I just published. It will also be included in the next release of pylance.
